src/dft.cpp

```diff
@@ -441,6 +441,151 @@ dft_flux fields::add_dft_flux(const volume_list *where_,
   return dft_flux(cE[0], cH[0], E, H, freq_min, freq_max, Nfreq);
 }
 
+dft_energy::dft_energy(dft_chunk *E_, dft_chunk *H_, 
+		       dft_chunk *D_, dft_chunk *B_, 
+		       double fmin, double fmax, int Nf)
+{
+  if (Nf <= 1) fmin = fmax = (fmin + fmax) * 0.5;
+  freq_min = fmin;
+  Nfreq = Nf;
+  dfreq = Nf <= 1 ? 0.0 : (fmax - fmin) / (Nf - 1);
+  E = E_; H = H_; D = D_; B = B_;
+}
+
+dft_energy::dft_energy(const dft_energy &f) {
+  freq_min = f.freq_min; Nfreq = f.Nfreq; dfreq = f.dfreq;
+  E = f.E; H = f.H; D = f.D; B = f.B;
+}
+
+double *dft_energy::electric() {
+  double *F = new double[Nfreq];
+  for (int i = 0; i < Nfreq; ++i) F[i] = 0;
+  for (dft_chunk *curE = E, *curD = D; curE && curD;
+       curE = curE->next_in_dft, curD = curD->next_in_dft)
+    for (int k = 0; k < curE->N; ++k)
+      for (int i = 0; i < Nfreq; ++i)
+	F[i] += 0.5*real(conj(curE->dft[k*Nfreq + i])
+			 * curD->dft[k*Nfreq + i]);
+  double *Fsum = new double[Nfreq];
+  sum_to_all(F, Fsum, Nfreq);
+  delete[] F;
+  return Fsum;
+}
+
+double *dft_energy::magnetic() {
+  double *F = new double[Nfreq];
+  for (int i = 0; i < Nfreq; ++i) F[i] = 0;
+  for (dft_chunk *curH = H, *curB = B; curH && curB;
+       curH = curH->next_in_dft, curB = curB->next_in_dft)
+    for (int k = 0; k < curH->N; ++k)
+      for (int i = 0; i < Nfreq; ++i)
+	F[i] += 0.5*real(conj(curH->dft[k*Nfreq + i])
+			 * curB->dft[k*Nfreq + i]);
+  double *Fsum = new double[Nfreq];
+  sum_to_all(F, Fsum, Nfreq);
+  delete[] F;
+  return Fsum;
+}
+
+double *dft_energy::total() {
+  double *Fe = electric();
+  double *Fm = magnetic();
+  double *F = new double[Nfreq];
+  for (int i = 0; i < Nfreq; ++i) F[i] = Fe[i]+Fm[i];
+  double *Fsum = new double[Nfreq];
+  sum_to_all(F, Fsum, Nfreq);
```

Comment:
This should not be necessary since there was already `sum_to_all` in `electric` and `magnetic`.